### ISSUE TITLE
plugin.managesieve Unable to connect

### ISSUE THREAD

**Issue by Spider84** [Mailu/Mailu#2762]
## Environment & Version

Debian 5.10.127-2 (2022-07-23) x86_64 GNU/Linux

### Environment

- [x] docker compose

### Version

- Version: `master`
- mailu/roundcube:1.9

## Description
Unable to open Sieve In Roundcube Setting.

## Replication Steps
RoundCube -> Settings -> Filter

## Logs

```

admin_1      | 49.237.9.236 - - [11/Apr/2023:14:49:46 +0600] "GET /internal/auth/user HTTP/1.0" 200 0 "https://mail.spider.vc/webmail/?_task=settings&_action=folders" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
imap_1       | Apr 11 14:49:49 managesieve-login: Error: haproxy: Client timed out (rip=192.168.203.10)
webmail_1    | errors: <04aqb1jj> PHP Error: Failed to read from socket (GET /?_task=settings&_action=plugin.managesieve)
webmail_1    | errors: <04aqb1jj> PHP Error: Unable to connect to managesieve on imap:4190 in /var/www/html/plugins/managesieve/lib/Roundcube/rcube_sieve_engine.php on line 227 (GET /?_task=settings&_action=plugin.managesieve)
webmail_1    | errors: <04aqb1jj> PHP Error: Not currently in AUTHORISATION state (GET /?_task=settings&_action=plugin.managesieve)
webmail_1    | errors: <04aqb1jj> PHP Error: Failed to write to socket: connection lost (GET /?_task=settings&_action=plugin.managesieve)
webmail_1    | 192.168.203.9 - - [11/Apr/2023:14:49:46 +0600] "GET /?_task=settings&_action=plugin.managesieve HTTP/1.1" 200 4766 "https://mail.spider.vc/webmail/?_task=settings&_action=folders" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
front_1      | 192.168.203.1 - - [11/Apr/2023:14:49:49 +0600] "GET /webmail/?_task=settings&_action=plugin.managesieve HTTP/1.0" 200 4191 "https://mail.spider.vc/webmail/?_task=settings&_action=folders" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
webmail_1    | fda0:fd3d:8:beef::a - - [11/Apr/2023:14:49:49 +0600] "GET /plugins/jqueryui/themes/larry/images/ui-dialog-close.png HTTP/1.1" 200 995 "https://mail.spider.vc/webmail/plugins/jqueryui/themes/larry/jquery-ui.min.css?s=1640816963" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
front_1      | 2023/04/11 14:49:49 [error] 15#15: *383651 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.203.1, server: , request: "GET /webmail/plugins/jqueryui/themes/larry/images/ui-dialog-close.png HTTP/1.0", subrequest: "/internal/auth/user", upstream: "http://[fda0:fd3d:8:beef::7]:80/internal/auth/user", host: "mail.spider.vc", referrer: "https://mail.spider.vc/webmail/plugins/jqueryui/themes/larry/jquery-ui.min.css?s=1640816963"
admin_1      | 49.237.9.236 - - [11/Apr/2023:14:49:49 +0600] "GET /internal/auth/user HTTP/1.0" 200 0 "https://mail.spider.vc/webmail/plugins/jqueryui/themes/larry/jquery-ui.min.css?s=1640816963" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
front_1      | 2023/04/11 14:49:49 [warn] 15#15: *383651 upstream server temporarily disabled while connecting to upstream, client: 192.168.203.1, server: , request: "GET /webmail/plugins/jqueryui/themes/larry/images/ui-dialog-close.png HTTP/1.0", subrequest: "/internal/auth/user", upstream: "http://[fda0:fd3d:8:beef::7]:80/internal/auth/user", host: "mail.spider.vc", referrer: "https://mail.spider.vc/webmail/plugins/jqueryui/themes/larry/jquery-ui.min.css?s=1640816963"
front_1      | 192.168.203.1 - - [11/Apr/2023:14:49:49 +0600] "GET /webmail/plugins/jqueryui/themes/larry/images/ui-dialog-close.png HTTP/1.0" 200 590 "https://mail.spider.vc/webmail/plugins/jqueryui/themes/larry/jquery-ui.min.css?s=1640816963" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
```

**Comment by nextgens** [comment]
Please clarify what version of the containers you are running (run ``cat /version`` in the container). Mixing up versions is obviously not supported.

**Comment by Spider84** [comment]
I has fixed version difference:
```
spider@mail:/home/spider/mailu$ docker-compose exec webmail cat /version
2.0.9
spider@mail:/home/spider/mailu$ docker-compose exec front cat /version
2.0.9
spider@mail:/home/spider/mailu$ docker-compose exec imap cat /version
2.0.9
```
but error not fixed:
```
front_1      | 172.25.0.3 [21/Apr/2023:14:44:56 +0600] TCP 502 0 0 0.001 "192.168.203.3:4190, [fda0:fd3d:8:beef::4]:4190" "0, 0" "0, 0" "-, -"
front_1      | 2023/04/21 14:44:56 [error] 17#17: *883 connect() failed (111: Connection refused) while connecting to upstream, client: 172.25.0.3, server: 0.0.0.0:14190, upstream: "192.168.203.3:4190", bytes from/to client:0/0, bytes from/to upstream:0/0
front_1      | 2023/04/21 14:44:56 [warn] 17#17: *883 upstream server temporarily disabled while connecting to upstream, client: 172.25.0.3, server: 0.0.0.0:14190, upstream: "192.168.203.3:4190", bytes from/to client:0/0, bytes from/to upstream:0/0
front_1      | 2023/04/21 14:44:56 [error] 17#17: *883 connect() failed (111: Connection refused) while connecting to upstream, client: 172.25.0.3, server: 0.0.0.0:14190, upstream: "[fda0:fd3d:8:beef::4]:4190", bytes from/to client:0/0, bytes from/to upstream:0/0
front_1      | 2023/04/21 14:44:56 [warn] 17#17: *883 upstream server temporarily disabled while connecting to upstream, client: 172.25.0.3, server: 0.0.0.0:14190, upstream: "[fda0:fd3d:8:beef::4]:4190", bytes from/to client:0/0, bytes from/to upstream:0/0
front_1      | 192.168.203.1 - - [21/Apr/2023:14:44:56 +0600] "GET /webmail/?_task=settings&_action=plugin.managesieve HTTP/1.0" 200 15706 "https://mail.xxx/webmail/?_task=settings&_action=plugin.managesieve" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36"
front_1      | 2023/04/21 14:44:56 [error] 17#17: *886 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.203.1, server: , request: "GET /static/mailu.png HTTP/1.0", upstream: "http://[fda0:fd3d:8:beef::7]:80/static/mailu.png", host: "mail.xxx", referrer: "https://mail.xxx/webmail/plugins/mailu/skins/elastic/mailu.css?s=1682010026"
front_1      | 2023/04/21 14:44:56 [warn] 17#17: *886 upstream server temporarily disabled while connecting to upstream, client: 192.168.203.1, server: , request: "GET /static/mailu.png HTTP/1.0", upstream: "http://[fda0:fd3d:8:beef::7]:80/static/mailu.png", host: "mail.xxx", referrer: "https://mail.xxx/webmail/plugins/mailu/skins/elastic/mailu.css?s=1682010026"
front_1      | 192.168.203.1 - - [21/Apr/2023:14:44:56 +0600] "GET /static/mailu.png HTTP/1.0" 304 0 "https://mail.xxx/webmail/plugins/mailu/skins/elastic/mailu.css?s=1682010026" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36"
```

**Comment by evasil** [comment]
Hello,

I am experiencing the same issue and same logs are on my setup.
I can see there is:

`managesieve-login: Error: haproxy: Client timed out`

So it is somehow related to haproxy, however I am sure I do not use haproxy anywhere in Mailu setup.
Or is haproxy a part of some container, eg. imap?

**Comment by nextgens** [comment]
> front_1      | 2023/04/21 14:44:56 [error] 17#17: *886 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.203.1, server: , request: "GET /static/mailu.png HTTP/1.0", upstream: "http://[fda0:fd3d:8:beef::7]:80/static/mailu.png", host: "mail.xxx", referrer: "https://mail.xxx/webmail/plugins/mailu/skins/elastic/mailu.css?s=1682010026"

This is fairly explicit: front has failed to reach admin over ipv6. Did you configure SUBNET6?
Those logs have nothing to do with sieve.

Connecting directly with sieveshell is expected to fail since it is not supposed to be exposed at all.

**Comment by Spider84** [comment]
sieve server of dovecot in imap docker configured as [haproxy](https://wiki.dovecot.org/HAProxy)
```
###############
# Filtering
###############
service managesieve-login {
  inet_listener sieve {
    port = 4190
    haproxy = yes
  }
}
```
then nginx in fron docker image is configured as 
```
stream {
    log_format main '$remote_addr [$time_local] '
                '$protocol $status $bytes_sent $bytes_received '
                '$session_time "$upstream_addr" '
                '"$upstream_bytes_sent" "$upstream_bytes_received" "$upstream_connect_time"';
    access_log /dev/stdout main;

    # managesieve
    server {
      listen 14190;
      resolver 127.0.0.11 valid=30s;

      proxy_connect_timeout 1s;
      proxy_timeout 1m;
      proxy_protocol on;
      proxy_pass imap:4190;
    }
}
```
Important past is `proxy_protocol on;` - this will force nginx to use haproxy extensions when proxying traffic to sieve. As I has understand - haproxy simple add client ip address information for next server (managesieve in our case).

In my case I got error connection refused - that mean managesieve droped connection attempt from nginx.

**Comment by Spider84** [comment]
> > front_1      | 2023/04/21 14:44:56 [error] 17#17: *886 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.203.1, server: , request: "GET /static/mailu.png HTTP/1.0", upstream: "http://[fda0:fd3d:8:beef::7]:80/static/mailu.png", host: "mail.xxx", referrer: "https://mail.xxx/webmail/plugins/mailu/skins/elastic/mailu.css?s=1682010026"
> 
> This is fairly explicit: front has failed to reach admin over ipv6. Did you configure SUBNET6? Those logs have nothing to do with sieve.

I haven't SUBNET6 configurations.
and 
```
front_1      | 172.25.0.3 [21/Apr/2023:14:44:56 +0600] TCP 502 0 0 0.001 "192.168.203.3:4190, [fda0:fd3d:8:beef::4]:4190" "0, 0" "0, 0" "-, -"
front_1      | 2023/04/21 14:44:56 [error] 17#17: *883 connect() failed (111: Connection refused) while connecting to upstream, client: 172.25.0.3, server: 0.0.0.0:14190, upstream: "192.168.203.3:4190", bytes from/to client:0/0, bytes from/to upstream:0/0
front_1      | 2023/04/21 14:44:56 [warn] 17#17: *883 upstream server temporarily disabled while connecting to upstream, client: 172.25.0.3, server: 0.0.0.0:14190, upstream: "192.168.203.3:4190", bytes from/to client:0/0, bytes from/to upstream:0/0
```
this was attempt of connection to sieve. But sieve has drop connection.

**Comment by nextgens** [comment]
It does look like ``imap`` is resolved to both ``192.168.203.3`` and ``fda0:fd3d:8:beef::4``. This will most likely break things.

Check whether sieve is actually bound/listening on imap; post the output of:
```
docker compose exec imap netstat -lnpt
```

**Comment by Spider84** [comment]
Any way if front resolve imap as two different addresses, it has tried connect to both. And as minimum "192.168.203.3" is right address for imap container. And connection by this address has failed and only then has tried connect to ipv6 address.
```
$ docker-compose exec imap netstat -lnpt                                                                                                                                                     Active Internet connections (only servers)                                                                                                                                                                       Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name                                                                                                                 tcp        0      0 0.0.0.0:2525            0.0.0.0:*               LISTEN      10/dovecot                                                                                                                       tcp        0      0 0.0.0.0:4190            0.0.0.0:*               LISTEN      10/dovecot                                                                                                                       tcp        0      0 0.0.0.0:993             0.0.0.0:*               LISTEN      10/dovecot                                                                                                                       tcp        0      0 0.0.0.0:995             0.0.0.0:*               LISTEN      10/dovecot                                                                                                                       tcp        0      0 127.0.0.11:34435        0.0.0.0:*               LISTEN      -                                                                                                                                tcp        0      0 0.0.0.0:5001            0.0.0.0:*               LISTEN      10/dovecot                                                                                                                       tcp        0      0 0.0.0.0:110             0.0.0.0:*               LISTEN      10/dovecot                                                                                                                       tcp        0      0 0.0.0.0:143             0.0.0.0:*               LISTEN      10/dovecot                                                                                                                       tcp        0      0 :::2525                 :::*                    LISTEN      10/dovecot                                                                                                                       tcp        0      0 :::4190                 :::*                    LISTEN      10/dovecot                                                                                                                       tcp        0      0 :::993                  :::*                    LISTEN      10/dovecot                                                                                                                       tcp        0      0 :::995                  :::*                    LISTEN      10/dovecot                                                                                                                       tcp        0      0 :::5001                 :::*                    LISTEN      10/dovecot                                                                                                                       tcp        0      0 :::110                  :::*                    LISTEN      10/dovecot                                                                                                                       tcp        0      0 :::143                  :::*                    LISTEN      10/dovecot
```
and
```
# docker-compose exec imap ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
1276: eth0@if1277: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:c0:a8:cb:07 brd ff:ff:ff:ff:ff:ff
    inet 192.168.203.3/24 brd 192.168.203.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fda0:fd3d:8:beef::4/64 scope global flags 02
       valid_lft forever preferred_lft forever
    inet6 fe80::42:c0ff:fea8:cb07/64 scope link
       valid_lft forever preferred_lft forever
```

**Comment by nextgens** [comment]
The expected answer was:
```
# docker compose exec imap netstat -lnpt 
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.11:46193        0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:2525            0.0.0.0:*               LISTEN      9/dovecot
tcp        0      0 0.0.0.0:993             0.0.0.0:*               LISTEN      9/dovecot
tcp        0      0 0.0.0.0:143             0.0.0.0:*               LISTEN      9/dovecot
tcp        0      0 0.0.0.0:5001            0.0.0.0:*               LISTEN      9/dovecot
tcp        0      0 0.0.0.0:4190            0.0.0.0:*               LISTEN      9/dovecot
tcp        0      0 :::2525                 :::*                    LISTEN      9/dovecot
tcp        0      0 :::993                  :::*                    LISTEN      9/dovecot
tcp        0      0 :::143                  :::*                    LISTEN      9/dovecot
tcp        0      0 :::5001                 :::*                    LISTEN      9/dovecot
tcp        0      0 :::4190                 :::*                    LISTEN      9/dovecot
```
If there is nothing bound, it obviously can't work. Please provide the logs of the ``imap`` container.
```
docker compose logs imap|tail -50
```

You may want to upgrade your docker compose too as the version you are running is not supported anymore.

**Comment by nextgens** [comment]
@evasil @Spider84 it may be easier to investigate/fix interactively; please join https://riot.im/app/#/room/#mailu:tedomum.net

**Comment by nextgens** [comment]
@Spider84 we've worked through @evasil 's issue, he had an old configuration option in his ``docker-compose.yml``.

Please confirm that you have rerun https://setup.mailu.io when you have upgraded or post its content here

**Comment by evasil** [comment]
@Spider84, 

that is now port 14190 which "front" is listening on now. My (and maybe yours) docker-compose.yaml is outdated.
Port 4190 does not need to be exposed on "imap" anymore.
Only port 14190 on "front", if you use external Webmail, or nothing if you use Webmail inside of the same docker scope.

**Comment by DBLaci** [comment]
I don't want to reopen this, but as this was the first issue which I found, maybe it could help others if I tell you what I did.

i upgrade mailu with the following steps:

```
docker-compose pull
docker-compose up -d
```

After one upgrade the sieve has broken.

I checked the docker-compose and .env if it matches with the recent generated - and it matched.

Before I tried to check logs or debug anything I did

```
docker-compose pull
docker-compose down
docker-compose up -d
```

And now it works. I guess upgrading without restarting sometimes break things. I don't know more, but if you have this error, maybe just start to restart the whole application - Windows style :D
